hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestDataNodeReconfiguration.java

```diff
@@ -756,4 +759,33 @@ public void testDfsUsageParameters() throws ReconfigurationException {
       }
     }
   }
+
+  public static class DummyCachingGetSpaceUsed extends CachingGetSpaceUsed {
+    public DummyCachingGetSpaceUsed(Builder builder) throws IOException {
+      super(builder.setInterval(1000).setJitter(0L));
+    }
+
+    @Override
+    protected void refresh() {
+      counter++;
+    }
+  }
+
+  @Test
+  public void testDfsUsageKlass() throws IOException, ReconfigurationException,
+          InterruptedException {
+
+    long lastCounter = counter;
+    Thread.sleep(5000);
+    assertTrue(counter > lastCounter);
+
+    for (int i = 0; i < NUM_DATA_NODE; i++) {
+      DataNode dn = cluster.getDataNodes().get(i);
+      dn.reconfigurePropertyImpl(FS_GETSPACEUSED_CLASSNAME, null);
+    }
+
+    lastCounter = counter;
+    Thread.sleep(5000);
+    assertEquals(lastCounter, counter);
+ }
```

Comment:
@singer-bin Thanks for updating the PR.
How about we stop using `conf.setClass` in `startDFCluster`, and reconfigure `DummyCachingGetSpaceUsed` in the unit test, and replace the order of the assertions? (But after this change, somehow the unit test failed.)
```suggestion
  @Test
  public void testDfsUsageKlass() throws IOException, ReconfigurationException,
          InterruptedException {

    long lastCounter = counter;
    Thread.sleep(5000);
    assertEquals(lastCounter, counter);

    for (int i = 0; i < NUM_DATA_NODE; i++) {
      DataNode dn = cluster.getDataNodes().get(i);
      dn.reconfigurePropertyImpl(FS_GETSPACEUSED_CLASSNAME, DummyCachingGetSpaceUsed.class.toString());
    }

    lastCounter = counter;
    Thread.sleep(5000);
    assertTrue(counter > lastCounter);
 }
```

Reply:
Hi, thanks for your comment. My UT ideas come from [HDFS-15120](https://issues.apache.org/jira/browse/HDFS-15120) ,you can take a look to help understand. My conf comes from the getNewConf() method as follows:

``` 
private void reconfSpaceUsedKlass(){ 
    List<FsVolumeImpl> volumeList = data.getVolumeList(); 
    for (FsVolumeImpl fsVolume : volumeList) { 
      Map<String, BlockPoolSlice> blockPoolSlices = fsVolume.getBlockPoolSlices(); 
      for (BlockPoolSlice bp : blockPoolSlices.values()) { 
        try { 
          bp.refreshSpaceUsedKlass(getNewConf()); 
        } catch (IOException e) { 
          e.printStackTrace(); 
        } 
      } 
    } 
  }
```

so , passing in `DummyCachingGetSpaceUsed.class.toString()` in

``` dn.reconfigurePropertyImpl(FS_GETSPACEUSED_CLASSNAME, DummyCachingGetSpaceUsed.class.toString()); ```
doesn't make sense to me.
I need to set it to my own custom class when the service starts, then wait for the counter to change, and then use zzz to refresh the value to the default.

Reply:
Looking forward to your comments and suggestions. @tasanuma

Reply:
@singer-bin 
- As well as other reconfiguring properties of DataNode, I prefer to be able to update parameters by calling `reconf.. .Parameters(property, newVal)`.
- I think this PR is similar to #3863. Instead of creating the new method of `reconfSpaceUsedKlass()`, how about extending `reconfDfsUsageParameters` to update `fs.getspaceused.classname`?

Reply:
Thank you very much for your comment, I will consider it. @tasanuma